<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ brentlabRnaSeqTools</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">brentlabRnaSeqTools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ProccessRun.html">Process a New Run From MGI</a>
    <a class="dropdown-item" href="../articles/QC_Library_Quality.html">QC: Library Quality</a>
    <a class="dropdown-item" href="../articles/QC_Sample_Agreement_Analysis_Tally.html">QC: Replicate Agreement Analysis and tally</a>
    <a class="dropdown-item" href="../articles/QC_Sample_Agreement_Objects.html">QC: Replicate Agreement Object Creation</a>
    <a class="dropdown-item" href="../articles/UsingRstudioProjects.html">UsingRstudioProjects</a>
    <a class="dropdown-item" href="../articles/envPerturbation.html">Environmental Perturbation Analysis Data Creation</a>
    <a class="dropdown-item" href="../articles/ninetyMinInduction_tally.html">90 Minute Induction Tally</a>
    <a class="dropdown-item" href="../articles/ninetyMinuteInduction_analysis.html">90 Minute Induction Analysis Data Creation</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="addQcColsToMeta.html">addQcColsToMeta()</a></code> 
        </dt>
        <dd>add qc metrics to metadata</dd>
      </dl><dl><dt>
          
          <code><a href="addRleSummaryStats.html">addRleSummaryStats()</a></code> 
        </dt>
        <dd>calculate RLE by replicate groups</dd>
      </dl><dl><dt>
          
          <code><a href="archiveDatabase.html">archiveDatabase()</a></code> 
        </dt>
        <dd>pull entire database (not counts) and save to output_dir for archival purposes</dd>
      </dl><dl><dt>
          
          <code><a href="autoAuditQcTable.html">autoAuditQcTable()</a></code> 
        </dt>
        <dd>Audit a qc table with metrics added</dd>
      </dl><dl><dt>
          
          <code><a href="brentlabRnaSeqExperiment.html">brentlabRnaSeqExperiment()</a></code> 
        </dt>
        <dd>brentlabRnaSeqExperiment constructor.</dd>
      </dl><dl><dt>
          
          <code><a href="brentlabRnaSeqSet.html">brentlabRnaSeqSet()</a></code> <code><a href="brentlabRnaSeqSet.html">brentlabRnaSeqSetFromDatabase()</a></code> 
        </dt>
        <dd>brentlabRnaSeqSet object and constructor</dd>
      </dl><dl><dt>
          
          <code><a href="brentlabRnaSeqSetTransform.html">brentlabRnaSeqSetTransform()</a></code> 
        </dt>
        <dd>brentlabRnaSeqSetTransform object and constructor. This extends DESeqTransform.</dd>
      </dl><dl><dt>
          
          <code><a href="calculateGeneWiseMedians.html">calculateGeneWiseMedians()</a></code> 
        </dt>
        <dd>calculate medians across rows of dataframe</dd>
      </dl><dl><dt>
          
          <code><a href="calculateRLE.html">calculateRLE()</a></code> 
        </dt>
        <dd>calculate RLE of a numeric dataframe</dd>
      </dl><dl><dt>
          
          <code><a href="coerceToDds.html">coerceToDds()</a></code> 
        </dt>
        <dd>coerce brentlabRnaSeqSet to DESeqDataSet</dd>
      </dl><dl><dt>
          
          <code><a href="connectToDatabase.html">connectToDatabase()</a></code> 
        </dt>
        <dd>Connect to a remote postgresql database</dd>
      </dl><dl><dt>
          
          <code><a href="createEnvPertSet_perturbed.html">createEnvPertSet_perturbed()</a></code> 
        </dt>
        <dd>create Environmental Perturbation Perturbed set</dd>
      </dl><dl><dt>
          
          <code><a href="createEnvPertSet_titrationWT.html">createEnvPertSet_titrationWT()</a></code> 
        </dt>
        <dd>create Environmental Perturbation Titration (TDY451 only) set</dd>
      </dl><dl><dt>
          
          <code><a href="createEnvPert_epWT.html">createEnvPert_epWT()</a></code> 
        </dt>
        <dd>filter combined_df for environmental perturbation sample set</dd>
      </dl><dl><dt>
          
          <code><a href="createExperimentSet.html">createExperimentSet()</a></code> 
        </dt>
        <dd>Create a pre-defined experiment set</dd>
      </dl><dl><dt>
          
          <code><a href="createIgvBatchscript.html">createIgvBatchscript()</a></code> 
        </dt>
        <dd>create IGV viewer batch script (single range, as many tracks as you like)</dd>
      </dl><dl><dt>
          
          <code><a href="createInductionSetTally.html">createInductionSetTally()</a></code> 
        </dt>
        <dd>create 90 minute induction set tally</dd>
      </dl><dl><dt>
          
          <code><a href="createNfCorePipelineSampleSheet.html">createNfCorePipelineSampleSheet()</a></code> 
        </dt>
        <dd>create nf-co sample sheet</dd>
      </dl><dl><dt>
          
          <code><a href="createNinetyMinuteInduction_2016grant.html">createNinetyMinuteInduction_2016grant()</a></code> 
        </dt>
        <dd>90minuteInduction set Definition</dd>
      </dl><dl><dt>
          
          <code><a href="createNinetyMinuteInduction_2016grantWithDoubles.html">createNinetyMinuteInduction_2016grantWithDoubles()</a></code> 
        </dt>
        <dd>ninetyMin set where both doubles are in the grant_df</dd>
      </dl><dl><dt>
          
          <code><a href="createNinetyMinuteInduction_all.html">createNinetyMinuteInduction_all()</a></code> 
        </dt>
        <dd>Samples with genotypes not in the grant_df</dd>
      </dl><dl><dt>
          
          <code><a href="createNinetyMinuteInduction_non2016grant.html">createNinetyMinuteInduction_non2016grant()</a></code> 
        </dt>
        <dd>Samples with genotypes not in the grant_df</dd>
      </dl><dl><dt>
          
          <code><a href="createNovoalignPipelineSamplesheet.html">createNovoalignPipelineSamplesheet()</a></code> 
        </dt>
        <dd>create a sample sheet for the novoalign pipeline</dd>
      </dl><dl><dt>
          
          <code><a href="createTestTrainSet.html">createTestTrainSet()</a></code> 
        </dt>
        <dd>create test train set</dd>
      </dl><dl><dt>
          
          <code><a href="database_colnames.html">database_colnames</a></code> 
        </dt>
        <dd>set global variables, mostly as a hack to make the tidyverse data binding pass the R CMD Check</dd>
      </dl><dl><dt>
          
          <code><a href="database_info.html">database_info</a></code> 
        </dt>
        <dd>URLS to active databases</dd>
      </dl><dl><dt>
          
          <code><a href="decomposeStatus2Bit.html">decomposeStatus2Bit()</a></code> 
        </dt>
        <dd>decompose sums of powers of two to a list of the summed powers</dd>
      </dl><dl><dt>
          
          <code><a href="estimateSizeFactorsByProtocol.html">estimateSizeFactorsByProtocol()</a></code> 
        </dt>
        <dd>Calculate Size Factors within protocol groups</dd>
      </dl><dl><dt>
          
          <code><a href="extractColData.html">extractColData()</a></code> 
        </dt>
        <dd>extract colData as tibble</dd>
      </dl><dl><dt>
          
          <code><a href="extractDesignMatrix.html">extractDesignMatrix()</a></code> 
        </dt>
        <dd>extract the model.matrix using the design and colData</dd>
      </dl><dl><dt>
          
          <code><a href="filterWtByExperimentalLibdate_90min.html">filterWtByExperimentalLibdate_90min()</a></code> 
        </dt>
        <dd>WT filtering helper for 90minInduction sets</dd>
      </dl><dl><dt>
          
          <code><a href="fltrLowReplicateParams.html">fltrLowReplicateParams()</a></code> 
        </dt>
        <dd>filter low replicate parameters from metadata</dd>
      </dl><dl><dt>
          
          <code><a href="geneGRanges.html">geneGRanges()</a></code> 
        </dt>
        <dd>Given a GenomicFeatures annotation_db and a gene_id, extract an GRanges
object of the cds</dd>
      </dl><dl><dt>
          
          <code><a href="getGeneNames.html">getGeneNames()</a></code> 
        </dt>
        <dd>Get gene names</dd>
      </dl><dl><dt>
          
          <code><a href="getMetadata.html">getMetadata()</a></code> 
        </dt>
        <dd>Get the combined metadata as a tibble from a remote database</dd>
      </dl><dl><dt>
          
          <code><a href="getRawCounts.html">getRawCounts()</a></code> 
        </dt>
        <dd>Get combined raw counts</dd>
      </dl><dl><dt>
          
          <code><a href="getRunNumberLeadingZero.html">getRunNumberLeadingZero()</a></code> 
        </dt>
        <dd>correct run number to add leading zero where approprirate</dd>
      </dl><dl><dt>
          
          <code><a href="getUserAuthToken.html">getUserAuthToken()</a></code> 
        </dt>
        <dd>get (via a http POST request) your user authentication token from the database</dd>
      </dl><dl><dt>
          
          <code><a href="grant_df.html">grant_df</a></code> 
        </dt>
        <dd>the 2016 grant summary represented as a dataframe</dd>
      </dl><dl><dt>
          
          <code><a href="htseq_libraryComplexity.html">htseq_libraryComplexity()</a></code> 
        </dt>
        <dd>Calculate percentage of library accounted for by top expressed genes</dd>
      </dl><dl><dt>
          
          <code><a href="htseq_locusLog2cpm.html">htseq_locusLog2cpm()</a></code> 
        </dt>
        <dd>Get the log2cpm of a given locus in the htseq output</dd>
      </dl><dl><dt>
          
          <code><a href="htseq_notAlignedTotalPercent.html">htseq_notAlignedTotalPercent()</a></code> 
        </dt>
        <dd>Extract notAlignedTotalPercent from novoalign log</dd>
      </dl><dl><dt>
          
          <code><a href="htseq_proteinCodingTotal.html">htseq_proteinCodingTotal()</a></code> 
        </dt>
        <dd>Calculate KN99 protein coding total</dd>
      </dl><dl><dt>
          
          <code><a href="isNumeric.html">isNumeric()</a></code> 
        </dt>
        <dd>test if argument is numeric</dd>
      </dl><dl><dt>
          
          <code><a href="is_integer64.html">is_integer64()</a></code> 
        </dt>
        <dd>Test if value is datatype integer64</dd>
      </dl><dl><dt>
          
          <code><a href="kn99_novo_htseq_status.html">kn99_novo_htseq_status</a></code> 
        </dt>
        <dd>QC status codes for the novo+htseq pipeline</dd>
      </dl><dl><dt>
          
          <code><a href="kn99_novo_htseq_thresholds.html">kn99_novo_htseq_thresholds</a></code> 
        </dt>
        <dd>thresholds set for the novo+htseq pipeline</dd>
      </dl><dl><dt>
          
          <code><a href="listTables.html">listTables()</a></code> 
        </dt>
        <dd>list tables in databse</dd>
      </dl><dl><dt>
          
          <code><a href="localView.html">localView()</a></code> 
        </dt>
        <dd>To view a data.frame / data.table in LibreOffice Calc</dd>
      </dl><dl><dt>
          
          <code><a href="locusCoverage.html">locusCoverage()</a></code> 
        </dt>
        <dd>Calculate coverage over a given feature</dd>
      </dl><dl><dt>
          
          <code><a href="moveNfCoFastqFiles.html">moveNfCoFastqFiles()</a></code> 
        </dt>
        <dd>move fastq files from lts to scratch for nf-co/rnaseq pipeline</dd>
      </dl><dl><dt>
          
          <code><a href="novoalignPipelineQC.html">novoalignPipelineQC()</a></code> 
        </dt>
        <dd>QC a crypto run output by the novoalign+htseq pipeline</dd>
      </dl><dl><dt>
          
          <code><a href="parseComparatives.html">parseComparatives()</a></code> 
        </dt>
        <dd>parse text comparative sentence</dd>
      </dl><dl><dt>
          
          <code><a href="parseNovoalignLog.html">parseNovoalignLog()</a></code> 
        </dt>
        <dd>Parse a novoalign log</dd>
      </dl><dl><dt>
          
          <code><a href="passing_genes_all.html">passing_genes_all</a></code> 
        </dt>
        <dd>genes considered NOT to have 'high dispersion'</dd>
      </dl><dl><dt>
          
          <code><a href="patchTable.html">patchTable()</a></code> 
        </dt>
        <dd>PATCH entries in database table</dd>
      </dl><dl><dt>
          
          <code><a href="postCounts.html">postCounts()</a></code> 
        </dt>
        <dd>post counts to database</dd>
      </dl><dl><dt>
          
          <code><a href="postFastqSheet.html">postFastqSheet()</a></code> 
        </dt>
        <dd>post new fastq sheet to database</dd>
      </dl><dl><dt>
          
          <code><a href="postQcSheet.html">postQcSheet()</a></code> 
        </dt>
        <dd>post new qc sheet to database</dd>
      </dl><dl><dt>
          
          <code><a href="postTable.html">postTable()</a></code> 
        </dt>
        <dd>Post a table to the database</dd>
      </dl><dl><dt>
          
          <code><a href="qaFilter.html">qaFilter()</a></code> 
        </dt>
        <dd>Quality filter a brentlabRnaSeqSet object</dd>
      </dl><dl><dt>
          
          <code><a href="readHTSeqFile.html">readHTSeqFile()</a></code> 
        </dt>
        <dd>Parse HTSeq output to tibble</dd>
      </dl><dl><dt>
          
          <code><a href="readInData.html">readInData()</a></code> 
        </dt>
        <dd>read in columnar data</dd>
      </dl><dl><dt>
          
          <code><a href="removeParameterEffects.html">removeParameterEffects()</a></code> 
        </dt>
        <dd>remove some effects from the counts</dd>
      </dl><dl><dt>
          
          <code><a href="replicateByProtocolTally.html">replicateByProtocolTally()</a></code> 
        </dt>
        <dd>experimental replicates by libraryProtocol tally</dd>
      </dl><dl><dt>
          
          <code><a href="replicateByProtocolTally_90min.html">replicateByProtocolTally_90min()</a></code> 
        </dt>
        <dd>90min replicate by protocol tally</dd>
      </dl><dl><dt>
          
          <code><a href="rleByReplicateGroup.html">rleByReplicateGroup()</a></code> 
        </dt>
        <dd>calculate RLE by replicate groups</dd>
      </dl><dl><dt>
          
          <code><a href="rlePlot.html">rlePlot()</a></code> 
        </dt>
        <dd>plot RLE for a given column filter</dd>
      </dl><dl><dt>
          
          <code><a href="rlePlotCompareEffectRemoved.html">rlePlotCompareEffectRemoved()</a></code> 
        </dt>
        <dd>plots output of rleSummaryByReplicateGroup</dd>
      </dl><dl><dt>
          
          <code><a href="rlePlot_helper.html">rlePlot_helper()</a></code> 
        </dt>
        <dd>the actual plotting function for rlePlot</dd>
      </dl><dl><dt>
          
          <code><a href="rleSummary.html">rleSummary()</a></code> 
        </dt>
        <dd>rleSummary calculates summary statistics of rleFullTable</dd>
      </dl><dl><dt>
          
          <code><a href="rleTransform.html">rleTransform()</a></code> 
        </dt>
        <dd>transform normalized counts into regularized log expression</dd>
      </dl><dl><dt>
          
          <code><a href="run_numbers_with_leading_zero.html">run_numbers_with_leading_zero</a></code> 
        </dt>
        <dd>A named list containing a run number without a leading zero, eg 647, with the value being the same runnumber with
a leading 0, eg 0647.</dd>
      </dl><dl><dt>
          
          <code><a href="sendCountsToDatabase.html">sendCountsToDatabase()</a></code> 
        </dt>
        <dd>Send Counts to Database (legacy pipeline)</dd>
      </dl><dl><dt>
          
          <code><a href="setPerturbedLociToZero.html">setPerturbedLociToZero()</a></code> 
        </dt>
        <dd>Zero out perturbed expression</dd>
      </dl><dl><dt>
          
          <code><a href="splitProtocolGroups.html">splitProtocolGroups()</a></code> 
        </dt>
        <dd>Split a brentlabRnaSeqSet into two by protocol</dd>
      </dl><dl><dt>
          
          <code><a href="strandedScanBamParam.html">strandedScanBamParam()</a></code> 
        </dt>
        <dd>A helper function to create a ScanBamParam object</dd>
      </dl><dl><dt>
          
          <code><a href="testBamPath.html">testBamPath()</a></code> 
        </dt>
        <dd>test bam path</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Chase Mateusiak.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer></div>

  

  

  </body></html>

