<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="brentlabRnaSeqTools">
<title>QC: Library Quality • brentlabRnaSeqTools</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="QC: Library Quality">
<meta property="og:description" content="brentlabRnaSeqTools">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">brentlabRnaSeqTools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ProccessRun.html">Process a New Run From MGI</a>
    <a class="dropdown-item" href="../articles/QC_Library_Quality.html">QC: Library Quality</a>
    <a class="dropdown-item" href="../articles/QC_Sample_Agreement_Objects.html">QC: Replicate Agreement</a>
    <a class="dropdown-item" href="../articles/Tally.html">Tally Experiment Sets</a>
    <a class="dropdown-item" href="../articles/UsingRstudioProjects.html">UsingRstudioProjects</a>
    <a class="dropdown-item" href="../articles/envPerturbation.html">Environmental Perturbation Analysis Data Creation</a>
    <a class="dropdown-item" href="../articles/ninetyMinuteInduction_analysis.html">90 Minute Induction Analysis Data Creation</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>QC: Library Quality</h1>
                        <h4 data-toc-skip class="author">Chase Mateusiak</h4>
            
            <h4 data-toc-skip class="date">02/09/2022</h4>
      
      
      <div class="d-none name"><code>QC_Library_Quality.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Details of library quality metrics and thresholds brentlabRnaSeqTools package version: 0.0.0.0</p>
    </div>
    
<div class="section level2">
<h2 id="cryptococcus">Cryptococcus<a class="anchor" aria-label="anchor" href="#cryptococcus"></a>
</h2>
<table class="table">
<thead><tr class="header">
<th>Metric</th>
<th>Threshold</th>
<th>Status</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Protein Coding Total</td>
<td>1e6</td>
<td>1</td>
</tr>
<tr class="even">
<td>Not Aligned Total Percent</td>
<td>.07</td>
<td>2</td>
</tr>
<tr class="odd">
<td>Perturbed Coverage</td>
<td>.25</td>
<td>4</td>
</tr>
<tr class="even">
<td>NAT coverage: expected</td>
<td>.5</td>
<td>8</td>
</tr>
<tr class="odd">
<td>NAT log2cpm: expected</td>
<td>5</td>
<td>8</td>
</tr>
<tr class="even">
<td>NAT coverage: unexpected</td>
<td>.5</td>
<td>16</td>
</tr>
<tr class="odd">
<td>NAT log2cpm: unexpected</td>
<td>2.5</td>
<td>16</td>
</tr>
<tr class="even">
<td>G418 log2cpm: expected</td>
<td>5.688</td>
<td>32</td>
</tr>
<tr class="odd">
<td>G418 log2cpm: unexpected</td>
<td>5.688</td>
<td>64</td>
</tr>
<tr class="even">
<td>overexpression FOW</td>
<td>2</td>
<td>128</td>
</tr>
<tr class="odd">
<td>missing marker in metadata</td>
<td></td>
<td>256</td>
</tr>
</tbody>
</table>
<p>Note that libraryComplexity is included in the qc metrics now, but there is no threshold currently set. The default setting for libraryComplexity is to calculate the portion of the total counts made up by the top 25 expressed genes</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">brentlabRnaSeqTools</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="get-the-metadata-from-the-database">Get the metadata from the database<a class="anchor" aria-label="anchor" href="#get-the-metadata-from-the-database"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meta</span> <span class="op">=</span> <span class="fu"><a href="../reference/getMetadata.html">getMetadata</a></span><span class="op">(</span>
  <span class="va">database_info</span><span class="op">$</span><span class="va">kn99</span><span class="op">$</span><span class="va">db_host</span>,
  <span class="va">database_info</span><span class="op">$</span><span class="va">kn99</span><span class="op">$</span><span class="va">db_name</span>,
  <span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"db_username"</span><span class="op">)</span>,
  <span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"db_password"</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="filter-out-the-reads-of-interest">Filter out the reads of interest<a class="anchor" aria-label="anchor" href="#filter-out-the-reads-of-interest"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">run_df</span> <span class="op">=</span> <span class="va">meta</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">runNumber</span> <span class="op">==</span> <span class="fl">5500</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-qc-table">Create QC table<a class="anchor" aria-label="anchor" href="#create-qc-table"></a>
</h2>
<p>You will need to have the cluster mounted to your local for this to work. Note that you may get warnings about the index file being older than the bam. That isn’t critical and can generally be ignored, though something I am aware of and trying to figure out.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pipeline_out</span> <span class="op">=</span> <span class="st">"/mnt/scratch/rnaseq_pipeline/align_count_results/run_5500/rnaseq_pipeline_results/run_5500_samples"</span>

<span class="va">run_qc</span> <span class="op">=</span> <span class="fu"><a href="../reference/novoalignPipelineQC.html">novoalignPipelineQC</a></span><span class="op">(</span><span class="va">run_df</span>, <span class="va">pipeline_out</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"kn99_stranded_gff_rds"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="parse-this-out-into-a-qc-sheet">parse this out into a QC sheet<a class="anchor" aria-label="anchor" href="#parse-this-out-into-a-qc-sheet"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">qc_table</span> <span class="op">=</span> <span class="fu"><a href="../reference/addQcColsToMeta.html">addQcColsToMeta</a></span><span class="op">(</span><span class="va">run_df</span>, <span class="va">run_qc</span><span class="op">)</span>


<span class="co"># write_csv(dplyr::select(qc_df, -c(genotype1, genotype2, marker1, marker2)), </span>
<span class="co">#           "~/Desktop/tmp/run_5500_qc.csv")</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="auto-and-manual-audit">auto and manual audit<a class="anchor" aria-label="anchor" href="#auto-and-manual-audit"></a>
</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">audited_qc_df</span> <span class="op">=</span> <span class="fu"><a href="../reference/autoAuditQcTable.html">autoAuditQcTable</a></span><span class="op">(</span><span class="va">qc_table</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="co"># add manual audit columns</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>manualAudit <span class="op">=</span> <span class="cn">NA</span>,
         manualStatus <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="update-manual-audit">update manual audit<a class="anchor" aria-label="anchor" href="#update-manual-audit"></a>
</h2>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">audited_qc_df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/edit.html" class="external-link">edit</a></span><span class="op">(</span><span class="va">audited_qc_df</span><span class="op">)</span>

<span class="va">audited_qc_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">genotype1</span>, <span class="va">genotype2</span>, <span class="va">marker1</span>, <span class="va">marker2</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
<span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_csv</a></span><span class="op">(</span><span class="st">"~/Desktop/tmp/run_5500_qc.csv"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="post-counts-to-database">post Counts to database<a class="anchor" aria-label="anchor" href="#post-counts-to-database"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">count_files</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.glob.html" class="external-link">Sys.glob</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pipeline_out</span>, <span class="st">"count"</span>, <span class="st">"*_read_count.tsv"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">count_files</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">count_files</span><span class="op">)</span>, <span class="st">"_read_count.tsv"</span><span class="op">)</span>

<span class="va">compiled_counts</span> <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">count_files</span><span class="op">)</span>, <span class="op">~</span><span class="fu"><a href="../reference/readHTSeqFile.html">readHTSeqFile</a></span><span class="op">(</span><span class="va">count_files</span><span class="op">[[</span><span class="va">.</span><span class="op">]</span><span class="op">]</span>, <span class="va">.</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu">plyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyr/man/join_all.html" class="external-link">join_all</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/startsWith.html" class="external-link">startsWith</a></span><span class="op">(</span><span class="va">feature</span>, <span class="st">"__"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">feature</span><span class="op">)</span>

<span class="va">fastq_df</span> <span class="op">=</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"/mnt/lts/sequence_data/rnaseq_data/kn99_database_archive/20220131/fastqFiles.csv"</span><span class="op">)</span>

<span class="co"># note: commented out to prevent me from accidently running this</span>
<span class="co"># res = postCounts(database_info$kn99$urls$counts, </span>
<span class="co">#                  5500, </span>
<span class="co">#                  Sys.getenv("kn99_db_token"), </span>
<span class="co">#                  compiled_counts, </span>
<span class="co">#                  fastq_df)</span>
<span class="co"># </span>
<span class="co"># res</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="send-qc-to-database">Send QC to database<a class="anchor" aria-label="anchor" href="#send-qc-to-database"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># note: commented out in the vignette to prevent me from running accidently</span>
<span class="co"># res = postQcSheet_test(database_info$kn99$urls$qualityAssess,</span>
<span class="co">#             Sys.getenv("kn99_db_token"),</span>
<span class="co">#             5500, </span>
<span class="co">#             "~/Desktop/tmp/run_5500_qc.csv",</span>
<span class="co">#             "/mnt/lts/sequence_data/rnaseq_data/kn99_database_archive/20220131/fastqFiles.csv")</span>

<span class="co"># a code of 200 or 201 means it worked. anything else means failure</span>
<span class="co"># res</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="igv-browser-shot">IGV browser shot<a class="anchor" aria-label="anchor" href="#igv-browser-shot"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">kn99_gff</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"kn99_stranded_gff_rds"</span><span class="op">)</span><span class="op">)</span>

<span class="va">unique_loci</span> <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">run_qc</span><span class="op">$</span><span class="va">locus</span><span class="op">)</span>, <span class="st">"CNAG"</span>, <span class="st">"CKF44"</span><span class="op">)</span>

<span class="va">unique_loci</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">unique_loci</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CNAG_NAT"</span>, <span class="st">"CNAG_G418"</span><span class="op">)</span><span class="op">)</span>

<span class="va">bam_prefix</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pipeline_out</span>, <span class="st">"align"</span><span class="op">)</span>
<span class="va">bam_suffix</span> <span class="op">=</span> <span class="st">"_sorted_aligned_reads_with_annote.bam"</span>

<span class="va">bam_list_df</span> <span class="op">=</span> <span class="va">run_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">fastqFileNumber</span>, .keep_all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="va">bam_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">bam_list_df</span>, <span class="va">fastqFileName</span><span class="op">)</span>, <span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">bam_prefix</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">.</span>, <span class="va">bam_suffix</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="va">bam_list</span> <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">bam_list</span>, <span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">.</span>, <span class="st">"/mnt/scratch/rnaseq_pipeline/align_count_results/run_5500/rnaseq_pipeline_results/run_5500_samples/align/Brent_3235_GTAC_1_SIC_Index2_08_TGAGGTTATC_AAGCACGT_S2_R1_001_sorted_aligned_reads_with_annote.bam"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bam_list</span><span class="op">)</span> <span class="op">=</span> <span class="va">bam_list_df</span><span class="op">$</span><span class="va">fastqFileNumber</span>

<span class="va">igvScriptAll</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">ffn</span><span class="op">)</span><span class="op">{</span>
  <span class="kw">for</span><span class="op">(</span><span class="va">locus</span> <span class="kw">in</span> <span class="va">unique_loci</span><span class="op">)</span><span class="op">{</span>
      <span class="va">granges</span> <span class="op">=</span> <span class="va">kn99_gff</span><span class="op">[</span><span class="va">kn99_gff</span><span class="op">$</span><span class="va">ID</span> <span class="op">==</span> <span class="va">locus</span> <span class="op">&amp;</span> <span class="va">kn99_gff</span><span class="op">$</span><span class="va">type</span> <span class="op">==</span> <span class="st">'gene'</span><span class="op">]</span>
      <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">granges</span><span class="op">)</span>
      <span class="va">basename</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">ffn</span>, <span class="va">locus</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>
      <span class="fu"><a href="../reference/createIgvBatchscript.html">createIgvBatchscript</a></span><span class="op">(</span>
        bam_list <span class="op">=</span> <span class="va">bam_list</span><span class="op">[[</span><span class="va">ffn</span><span class="op">]</span><span class="op">]</span>,
        granges <span class="op">=</span> <span class="va">granges</span>,
        igv_genome <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"kn99_stranded_igv_genome"</span><span class="op">)</span>,
        output_dir <span class="op">=</span> <span class="st">"/home/oguzkhan/Desktop/tmp/igv/"</span>,
        output_file_basename <span class="op">=</span> <span class="va">basename</span><span class="op">)</span>
  <span class="op">}</span>

<span class="op">}</span>

<span class="fu">igvScriptAll</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bam_list</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="running-the-igv-batch-scripts">Running the IGV batch scripts<a class="anchor" aria-label="anchor" href="#running-the-igv-batch-scripts"></a>
</h2>
<p>It is likely easiest just to cd into the place where you output the scripts</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /home/oguzkhan/Desktop/tmp/igv/scripts</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> batch_script <span class="kw">in</span> <span class="va">$(</span><span class="fu">ls</span> .<span class="va">)</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="ex">xvfb-run</span> <span class="at">--auto-servernum</span> igv.sh <span class="at">-b</span> batch_script</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code></pre></div>
<p>I then scp these into the run directory before moving the run directory over to <code>lts</code></p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scp</span> <span class="at">-r</span> /home/oguzkhan/Desktop/tmp/igv chasem@htcf.wustl.edu:/mnt/scratch/rnaseq_pipeline/align_count_results/run_5500/rnaseq_pipeline_results/run_5500_samples/</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="co"># then log into htcf, go to the pipeline output and move the whole run directory to lts</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rsync</span> <span class="at">-aHv</span> run_5500_samples /lts/mblab/sequence_data/rnaseq_data/lts_align_count/</span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Chase Mateusiak.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
